rest-api-model-ql/src/main/kotlin/org/modelix/sample/restapimodelql/LightModelClientWrapper.kt

```diff
@@ -0,0 +1,120 @@
+
+        // the modelQL query
+        this.lightModelClient.changeQuery(buildModelQuery {
+            // we traverse from the root (a repository)
+            root {
+                // to all modules
+                children("modules") {
+                    // selecting the ones that contain our desired model name
+                    whereProperty("name").contains(MPS_MODEL_NAME)
+                    children("models") {
+                        // and extract all root nodes, so Rooms and Courses, and their descendants
+                        children("rootNodes") {
+                            descendants { }
+                        }
+                    }
+                }
+            }
+        })
+        logger.info("Connection successful")
+    }
+
+    suspend fun getAllLectures(): List<N_Lecture> {
+        //TODO: can we avoid this cast by using generics in runAsyncRead?
+        logger.info("Loading all lectures")
+        return runAsyncRead(loadLectures) as List<N_Lecture>
+    }
+
+    suspend fun getAllRooms(): List<N_Room> {
+        logger.info("Loading all rooms")
+        return runAsyncRead(loadRooms) as List<N_Room>
+    }
+
+    private suspend fun runAsyncRead(givenFunction: (node: N_Repository) -> List<N_BaseConcept>?): List<N_BaseConcept>? {
+        var result: List<N_BaseConcept>? = null
+        kotlinx.coroutines.withTimeout(5.seconds) {
+            while (true) {
+                lightModelClient.checkException()
+                val node = lightModelClient.runRead { lightModelClient.getRootNode() }
+                if (node != null && lightModelClient.runRead { node.isValid }) {
+                    lightModelClient.runRead {
+                        logger.info("Reading repository")
+                        result = givenFunction(node.typed<N_Repository>())
+                    }
+                    break
+                }
+                delay(10.milliseconds)
+            }
+        }
+        return result
+    }
+
+    private var loadRooms: (N_Repository) -> List<N_Room>? = Any@{ node: N_Repository ->
+        return@Any getAllRootNodesOfTypeInRepository<N_Rooms>(node).rooms
+    }
+
+    private var loadLectures: (N_Repository) -> List<N_Lecture>? = Any@{ node: N_Repository ->
+        return@Any getAllRootNodesOfTypeInRepository<N_Courses>(node).lectures
+    }
+
+    private inline fun <reified R> getAllRootNodesOfTypeInRepository(node: N_Repository) =
+            node.unwrap().allChildren.filter { it.isValid }.map { it.typed<N_Module>() }.models.rootNodes.filterIsInstance<R>()
+
+    val resolveNodeIdToConcept: suspend (String) -> N_BaseConcept? = Any@{ ref: String ->
+        logger.info("Resolving node $ref")
+        return@Any lightModelClient.runRead {
+            if (lightModelClient.getNodeIfLoaded(ref)?.typed() != null) {
+                lightModelClient.getNodeIfLoaded(ref)?.typed() as N_BaseConcept
+            } else {
+                null
+            }
```

Comment:
```suggestion
            lightModelClient.getNodeIfLoaded(ref)?.typed()?.let { it  as N_BaseConcept }
```

Would be an alternative way in Kotlin to write this.